spec/lang/representation.md

```diff
@@ -131,25 +131,28 @@ fn encode_ptr<M: Memory>(ptr: ThinPointer<M::Provenance>) -> List<AbstractByte<M
 }
 
 impl PointerMetaKind {
-    /// Returns the representable type of the metadata if there is any.
-    pub fn ty<T: Target>(self) -> Option<Type> {
+    /// Returns the representable type of the metadata.
+    pub fn ty<T: Target>(self) -> Type {
         match self {
-            PointerMetaKind::None => None,
-            PointerMetaKind::ElementCount => Some(Type::Int(IntType::usize_ty::<T>())),
-            PointerMetaKind::VTablePointer(..) => Some(Type::Ptr(PtrType::VTablePtr)),
+            PointerMetaKind::None => unit_type(),
+            PointerMetaKind::ElementCount => Type::Int(IntType::usize_ty::<T>()),
+            PointerMetaKind::VTablePointer(..) => Type::Ptr(PtrType::VTablePtr),
         }
     }
 }
 
 impl PtrType {
     /// Returns a pair type representing this wide pointer or `None` if it is thin.
     pub fn as_wide_pair<T: Target>(self) -> Option<Type> {
-        let meta_ty = self.meta_kind().ty::<T>()?;
+        if self.meta_kind() == PointerMetaKind::None {
+            return None;
+        }
+        let meta_ty = self.meta_kind().ty::<T>();
         let thin_pointer_field = (Offset::ZERO, Type::Ptr(PtrType::Raw { meta_kind: PointerMetaKind::None }));
         let metadata_field = (T::PTR_SIZE, meta_ty);
         ret(Type::Tuple {
             fields: list![thin_pointer_field, metadata_field],
-            size: Int::from(2) * T::PTR_SIZE,
+            size: T::PTR_SIZE + meta_ty.layout::<T>().expect_size("metadata is always sized"),
```

Comment:
This is still making some subtle assumptions about alignment. So probably we should just assert that the metadata has the same size and alignment as a pointer, and then `2 * PTR_SIZE` makes perfect sense here.